File: Data Structures & Algorithms/permutations/submission-0.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []

        perm = []
        pick = [False] * len(nums)
        def dfs():
            if len(perm) == len(nums):
                res.append(perm.copy())
                return
            for i,num in enumerate(nums):
                if not pick[i]:
                    perm.append(num)
                    pick[i] = True
                    dfs()
                    perm.pop()
                    pick[i] = False
        dfs()
        return res

File: Data Structures & Algorithms/count-subsequences/submission-0.cpp
class Solution {
public:
  int numDistinct(string s, string t) {
    memo.resize(s.size(), vector<int>(t.size(), -1));
    return dfs(s, t, 0, 0);
  }
  int dfs(const string s, const string t, int i, int j) {
    if (j == t.size()) {
      return 1;
    }
    if (i == s.size() && j < t.size())
      return 0;
    if (memo[i][j] != -1) {
      return memo[i][j];
    }
    if (s[i] == t[j]) {
      return memo[i][j] = dfs(s, t, i + 1, j + 1) + dfs(s, t, i + 1, j);
    } else {
      return memo[i][j] = dfs(s, t, i + 1, j);
    }
  }
private:
  vector<vector<int>> memo;

};
